datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1335,39 +1335,16 @@ private void handleLeaderDoAssignment(boolean cleanUpOrphanNodes) {
       // and schedule a retry. The retry should be able to diff the remaining ZooKeeper work
       if (_config.getEnableAssignmentTokens()) {
         _adapter.updateAllAssignmentsAndIssueTokens(newAssignmentsByInstance, stoppingDatastreamGroups);
+        try {
+          _tokenClaimExecutor.submit(() -> waitForStopToPropagateAndMarkDatastreamsStopped(stoppingDatastreamGroups));
+        } catch (RejectedExecutionException ex) {
+          _log.error("handleLeaderDoAssignment: Failed to submit the task that waits for stop propagation");
+          succeeded = false;
+        }
+
       } else {
         _adapter.updateAllAssignments(newAssignmentsByInstance);
-      }
-
-      Set<String> failedStreams = Collections.emptySet();
-      // Poll the zookeeper to ensure that hosts claimed assignment tokens for stopping streams
-      if (_config.getEnableAssignmentTokens() &&
-          !PollUtils.poll(() -> _adapter.getNumUnclaimedTokensForDatastreams(stoppingDatastreamGroups) == 0,
-            STOP_PROPAGATION_RETRY_MS, _config.getStopPropagationTimeout())) {
-        Map<String, List<AssignmentToken>> unclaimedTokens =
-            _adapter.getUnclaimedAssignmentTokensForDatastreams(stoppingDatastreamGroups);
-        failedStreams = unclaimedTokens.keySet();
-        Set<String> hosts = unclaimedTokens.values().stream().flatMap(List::stream).
-            map(AssignmentToken::getIssuedFor).collect(Collectors.toSet());
-        _log.error("Stop failed to propagate within {}ms for streams: {}. The following hosts failed to claim their token(s): {}",
-            _config.getStopPropagationTimeout(), failedStreams, hosts);
-        revokeUnclaimedAssignmentTokens(unclaimedTokens, stoppingDatastreamGroups);
-        _metrics.updateMeter(CoordinatorMetrics.Meter.NUM_FAILED_STOPS, failedStreams.size());
-      }
-
-      // TODO Explore introduction of UNKNOWN/WARN state for streams that failed to stop
-
-      for (DatastreamGroup datastreamGroup : stoppingDatastreamGroups) {
-        for (Datastream datastream : datastreamGroup.getDatastreams()) {
-          // Only streams that were confirmed to have stopped successfully will be transitioned to STOPPED state
-          if (!failedStreams.contains(datastream.getName())) {
-            datastream.setStatus(DatastreamStatus.STOPPED);
-            if (!_adapter.updateDatastream(datastream)) {
-              _log.warn("Failed to update datastream: {} to stopped state", datastream.getName());
-              succeeded = false;
-            }
-          }
-        }
+        succeeded = markDatastreamsStopped(stoppingDatastreamGroups, Collections.emptySet());
```

Comment:
I left the succeeded flag assignment here to not introduce regressions when `enableAssignmentTokens` is disabled (i.e. to make sure the behavior is unchanged)

Reply:
(a) If succeed flag is true, then `cleanUpDeadInstanceDataAndOtherUnusedTasks` removes the dead instances from zk.
(b) ` _adapter.updateAllAssignments(newAssignmentsByInstance)` save the newAssignment in the cache and get rid of the older assignment.

Both operations should happen atomically. But, there is a bug in this path introduced because of the datastream stopping transition; if the datastream transition to stopped failed, then only (b) will happen. This will result in the next leader seeing stale assignments and going into a non-recoverable state.